### DISCUSSION TITLE
Autocomplete select field with large dataset

### DISCUSSION THREAD

**Discussion by diogenesbrussels** [Budibase/budibase#10258] [Help] [answered]
Hello everyone,
I have a problem with the use of select fields and the autocomplete on large datasets.
In my forms, I have many times fields that ask the user to choose an item in a large list (5000 to 10000 rows).
Autocomplete does not work for these large lists because it seems to me that it works like this: the dataset is requested and then the autocomplete filter is applied on this dataset. However, in the case of a large dataset, only the first elements (1000?) are returned, which makes the search non-functional.
I have implemented the following solution but it makes the use by my users complicated: a text field allows a first filter and when a change is detected the dataset is updated and the select field contains less values. I have received complaints because it is difficult on mobile and requires a lot of clicks.

Here are the solutions I have considered:
- Modify the Select component so that autocomplete is based on the first characters and updates the dataset according to them. I don't know if this is desirable or possible.
- Create a search form on a separate page that contains search fields, a repeater with a select button for each item and then display it modally after clicking a button, then update the value in my form after the user selects.
- Create a Side Panel with the same selection and search form as in the previous proposal.

My constraints are the following:
- These select fields with large dataset are multiple: choice of a place, an organisation, etc. and can be present several times per form.
- The autocomplete filter does not work well for searches in languages other than English (with accents for example) and does not allow searches in several columns, so I use the full-text search function of MySQL
- I do not use relational links between these data in Budibase because it is not possible to maintain multiple relationships to the same table (#3854)

Would you have a suggestion on the best way to solve this problem? The first solution seems more universal to me as the other two require creating many elements and adapting them for each field?

Thank you in advance for your feedback and possible solutions!
Olivier

**Comment by melohagan** [comment]
Hey @diogenesbrussels 

> However, in the case of a large dataset, only the first elements (1000?) are returned

This is correct:

```js
/** @description The maximum number of rows to return, useful when paginating, for internal tables this will be limited to 1000, for SQL tables it will be 5000. */
limit?: number
```

The **autocomplete** on an Options Picker is simply a front-end JavaScript filter, and thus will be limited by the initial 100 records returned:
https://github.com/Budibase/budibase/blob/develop/packages/bbui/src/Form/Core/Picker.svelte#L79

---

So instead of using the options picker which has a limited filter, I would go with the repeater approach as this will make full use of the underlying filter capabilities.
 > Create a search form on a separate page that contains search fields, a repeater with a select button for each item and then display it modally after clicking a button, then update the value in my form after the user selects. 

You don't necessarily need a separate page; although it may present better in a separate screen modal. You could also use the Card block for convenience here.
The card block can have the filter applied, and then on click of a particular card title, you can use the **Update Field Value** action to update an existing read-only text field on your form to be submitted.

**Reply by diogenesbrussels** [reply]
Hey @melohagan 

Thank you for the information and clarifications!

I'll start on a separate page as this type of choice occurs in multiple places and including a copy of it in every form will make maintenance very complex in the future. This allows me to have one page for each type of search to maintain and use it wherever needed.

I still have questions about the best way to implement this though:
- When the user clicks on the button to choose, I open a modal window, pass in the name of the field that should contain the returned value (via url?)
- Once the user has chosen their item in the modal window, clicking the button should close the window and change the field value in the base form

What is the best practice for returning this value to the correct field? The modal window does not have access to the context of the form page, how do you do this cleanly? I could have forms with several of these 'modal choices'.

Thanks for your feedback!

**Reply by melohagan** [reply]
Hey @diogenesbrussels 

I had actually done the work to load values from a modal into the parent screen here: https://github.com/Budibase/budibase/pull/8016

If you scroll down to **Reload same screen** you can see how to pass variables from the URL into the parent screen. A reload is then triggered so that *Default values* in your form will reload!

**Reply by melohagan** [reply]
Sorry I should have read the conversations, I forgot I had rolled that back: https://github.com/Budibase/budibase/pull/8016#issuecomment-1262059015

Would you be able to raise a feature request issue for this scenario? I don't believe there's an issue out there at the moment.

**Reply by diogenesbrussels** [reply]
Hey @melohagan 

Sure, I will create this, should it be around ?
- sending a result from a modal window to the parent (on closing?)
- ability for a modal window to modify the fields of its parent
- ability to detect the change of a State Value in the parent when it is changed in the modal window

If you can think of a way to do this in the meantime, I'd love to hear from you :)

Thanks!

**Reply by melohagan** [reply, marked as answer]
A feature request for the first two points would be perfect:

> - sending a result from a modal window to the parent (on closing?)
> - ability for a modal window to modify the fields of its parent

For the final point it should already be possible to update state in a modal and access that in the parent screen. 

The problem currently is that the default values of a form field only display on load of the screen, thus even if the value is tied to state you still need to refresh the screen.

You also cannot update the form fields of the parent screen from the modal with the **Update Field Value** action. This issue relates to the feature request.

In the meantime, the solutions that come to mind might not be ideal:

1. Can you use [Side panels](https://docs.budibase.com/docs/side-panel)?

2. You could create a [custom component](https://docs.budibase.com/docs/custom-component) that includes a button that opens a modal on click. The component would need to support nested children as the content of your modal (which could be form fields). Then because it's in the same screen, the custom modal would have access to the context of the parent screen.

3. In order to get the default values to trigger from the loading of a screen you could have two identical screens. Then when the parent is one screen, closing the modal could navigate to the other screen and vice versa. It's a bit of a hack, but it would force the refresh of default values.

**Reply by diogenesbrussels** [reply]
Ok, thank you!

I am currently testing the solution with a side panel that contains the necessary information to search and select an item.

However, I am encountering a name collision bug between the names of the original form and the side panel search form. I will add a bug report for this specifically. The behaviour is not what I expected but I will investigate to find out why.

I'll keep you posted on the solution that works.

**Reply by diogenesbrussels** [reply]
Feature request here: #10399

**Reply by diogenesbrussels** [reply]
I have finally almost completed an implementation of a solution that works as follows:
- A button that opens a side-panel
- This side-panel contains a search form that uses a data provider and a repeater to display the search results
- Each result contains a button that closes the side panel and changes the value in a hidden field of the main form
- When the hidden field is updated, a data provider in the main form will take the information from the selected item to display the user's choice

This seems to work but : 
- with a specific solution for some of the fields in the sidebar search form (as it is a form within a form) because when one of these fields has the same name as a field in the main form it takes the value from the parent form! If the field is updated, it is the correct field and therefore the correct value that is taken into account. So I used States for these fields (this adds some unnecessary complications)
- I have this type of need sometimes 4 times per form, which adds as many side-panels, sub-forms, data-providers, etc. With the add and modify form, that's 8 components to maintain for a single data type.
- I had to hide the data providers from the search results because even when the side-panel is hidden it still performs the query (sometimes 2 or 3 times - multiplied by the number of side-panels) which greatly impacts the loading performance
